src/lib/expression/evaluation/expression_evaluator.cpp

```diff
@@ -253,27 +254,97 @@ ExpressionEvaluator::_evaluate_in_expression<ExpressionEvaluator::Bool>(const In
-        type_compatible_elements.emplace_back(element);
+    bool all_elements_are_values_of_left_type = true;
+    resolve_data_type(left_expression.data_type(), [&](const auto left_data_type_t) {
+      using LeftDataType = typename decltype(left_data_type_t)::type;
+
+      for (const auto& element : array_expression.elements()) {
+        if ((element->data_type() == DataType::String) == left_is_string) {
+          type_compatible_elements.emplace_back(element);
+        }
+
+        if (element->type != ExpressionType::Value) {
+          all_elements_are_values_of_left_type = false;
+        } else {
+          const auto& value_expression = std::static_pointer_cast<ValueExpression>(element);
+          if (value_expression->value.type() != typeid(LeftDataType)) all_elements_are_values_of_left_type = false;
+        }
       }
-    }
+    });
 
     if (type_compatible_elements.empty()) {
-      // `5 IN ()` is FALSE as is `NULL IN ()`
+      // `5 IN ()` is FALSE
       return std::make_shared<ExpressionResult<ExpressionEvaluator::Bool>>(std::vector<ExpressionEvaluator::Bool>{0});
     }
 
+    // If all elements of the list are simple values (e.g., `IN (1, 2, 3)`), iterate over the column and compare the
+    // values. Otherwise, translate the IN clause to a series of ORs:
+    // "a IN (x, y, z)"   ---->   "a = x OR a = y OR a = z"
+    if (all_elements_are_values_of_left_type) {
+      _resolve_to_expression_result_view(left_expression, [&](const auto& view) {
+        using LeftDataType = typename std::decay_t<decltype(view)>::Type;
+
+        // Above, we have ruled out NULL on the left side, but the compiler does not know this yet
+        if constexpr (!std::is_same_v<LeftDataType, NullValue>) {
+          std::vector<LeftDataType> right_values;
+          right_values.reserve(type_compatible_elements.size());
+          for (const auto& expression : type_compatible_elements) {
+            const auto& value_expression = std::static_pointer_cast<ValueExpression>(expression);
+            right_values.emplace_back(boost::get<LeftDataType>(value_expression->value));
+          }
+          std::sort(right_values.begin(), right_values.end());
+
+          result_values.resize(view.size());
+          if (left_expression.is_nullable()) {
+            result_nulls.resize(view.size());
+            for (auto chunk_offset = ChunkOffset{0}; chunk_offset < view.size(); ++chunk_offset) {
+              if (view.is_null(chunk_offset)) {
+                result_nulls[chunk_offset] = true;
+                continue;
+              }
+              if (auto it = std::lower_bound(right_values.cbegin(), right_values.cend(), view.value(chunk_offset));
+                  it != right_values.cend() && *it == view.value(chunk_offset)) {
+                result_values[chunk_offset] = true;
+              }
+            }
+          } else {
+            for (auto chunk_offset = ChunkOffset{0}; chunk_offset < view.size(); ++chunk_offset) {
```

Comment:
pls de-duplicate, you can fill the value- and null vectors separately while checking for `view.is_null(chunk_offset)` before doing the binary search to avoid doing it if the left operand is null.

Reply:
I can't follow. If I fill the vectors separately, won't I have to iterate over the view twice?

Reply:
Ok, so I would first fill the null values, then the "real" values. When I do the second part, won't I have to iterate over the null vector again to know which values to skip?

Reply:
Yeah, I think that's what I meant. Just to make things clear: 
1. Generate the result_nulls vector from `view.is_null(chunk_offset)`. Make sure it has size() == 0 if the view is not nullable.
2. Generate the result_values, but before computing it for a chunk_offset, first check whether the ->view<- is null at a particular row. Make sure you check this on the view and not the result_nulls vector because, remember, the result_nulls vector is empty if the view isn't nullable.

Reply:
This sounds like calling `view.is_null` twice, i.e., like three vector iterations instead of two

Reply:
Yep, doing it as described in my previous reply would mean iterating the input null vector twice, hadn't thought about that.

You can work around this.
Before entering the loop, allocate the result_values vector. Also, allocate the result_nulls vector, but only if `view.is_nullable()`.
Populate them in the same loop, but only write to result_nulls if the `view.is_nullable()` - this will be known at compile time so there is no overhead checking it on every iteration. Only compute a value for result_values[chunk_offset] if `!view.is_null(chunk_offset)`

Reply:
> Populate them in the same loop, but only write to result_nulls if the view.is_nullable() - this will be known at compile time so there is no overhead checking it on every iteration.

`bool is_nullable() const { return !_nulls.empty(); }` does not look like it could be a `constexpr`.

Reply:
Uups, that's an error in `ExpressionResultNullableSeries` then, that should always return true. Lemme check

Reply:
Wait a second. Even if `is_nullable()` is `constexpr`, it doesn't help us. `select_expression.is_nullable()` cannot be decided at compile time because we have no idea what `select_expression` is.

There is the possibility to do this with a helper method with `template<bool>`, but I don't think that it improves the code quality, which was the reason for this comment.

Reply:
I searched the code for `select_expression.is_nullable()` but couldn't find it - which file/line are you referring to? 

(...and naturally `select_expression.is_nullable()` cannot be decided at compile time. That is why you resolve the select_expression to a view. The nullability for View types is fixed at runtime - except, as you noted elsewhere - for Literal types.)